File: brainchain/mcp/servers.py
"""
Built-in MCP server configurations.

Provides pre-configured server settings for common MCP servers.
"""

from dataclasses import dataclass, field
from typing import Any


@dataclass
class MCPServerConfig:
    """Configuration for an MCP server."""
    command: list[str]
    name: str = ""
    args: dict[str, Any] = field(default_factory=dict)
    env: dict[str, str] | None = None
    enabled: bool = True
    auto_connect: bool = False
    timeout: int = 30

    def with_args(self, **kwargs) -> "MCPServerConfig":
        """Create a copy with updated args."""
        new_args = {**self.args, **kwargs}
        return MCPServerConfig(
            command=self.command.copy(),
            name=self.name,
            args=new_args,
            env=self.env.copy() if self.env else None,
            enabled=self.enabled,
            auto_connect=self.auto_connect,
            timeout=self.timeout,
        )


# Built-in server configurations
BUILTIN_SERVERS: dict[str, MCPServerConfig] = {
    "filesystem": MCPServerConfig(
        name="filesystem",
        command=["npx", "-y", "@modelcontextprotocol/server-filesystem"],
        args={"allowed_paths": ["."]},
        enabled=True,
    ),
    "fetch": MCPServerConfig(
        name="fetch",
        command=["npx", "-y", "@modelcontextprotocol/server-fetch"],
        args={},
        enabled=False,
    ),
    "memory": MCPServerConfig(
        name="memory",
        command=["npx", "-y", "@modelcontextprotocol/server-memory"],
        args={},
        enabled=False,
    ),
    "puppeteer": MCPServerConfig(
        name="puppeteer",
        command=["npx", "-y", "@modelcontextprotocol/server-puppeteer"],
        args={},
        enabled=False,
    ),
    "brave-search": MCPServerConfig(
        name="brave-search",
        command=["npx", "-y", "@modelcontextprotocol/server-brave-search"],
        args={},
        env={"BRAVE_API_KEY": ""},
        enabled=False,
    ),
    "github": MCPServerConfig(
        name="github",
        command=["npx", "-y", "@modelcontextprotocol/server-github"],
        args={},
        env={"GITHUB_PERSONAL_ACCESS_TOKEN": ""},
        enabled=False,
    ),
    "postgres": MCPServerConfig(
        name="postgres",
        command=["npx", "-y", "@modelcontextprotocol/server-postgres"],
        args={"connection_string": ""},
        enabled=False,
    ),
    "sqlite": MCPServerConfig(
        name="sqlite",
        command=["npx", "-y", "@modelcontextprotocol/server-sqlite"],
        args={"database_path": ""},
        enabled=False,
    ),
    "slack": MCPServerConfig(
        name="slack",
        command=["npx", "-y", "@modelcontextprotocol/server-slack"],
        args={},
        env={"SLACK_BOT_TOKEN": "", "SLACK_TEAM_ID": ""},
        enabled=False,
    ),
    "google-maps": MCPServerConfig(
        name="google-maps",
        command=["npx", "-y", "@modelcontextprotocol/server-google-maps"],
        args={},
        env={"GOOGLE_MAPS_API_KEY": ""},
        enabled=False,
    ),
}


def get_server_config(name: str) -> MCPServerConfig | None:
    """
    Get a built-in server configuration by name.

    Args:
        name: Server name

    Returns:
        MCPServerConfig or None if not found
    """
    return BUILTIN_SERVERS.get(name)


def list_builtin_servers() -> list[str]:
    """
    List names of all built-in servers.

    Returns:
        List of server names
    """
    return list(BUILTIN_SERVERS.keys())


def create_server_config(
    command: list[str],
    name: str = "custom",
    **kwargs,
) -> MCPServerConfig:
    """
    Create a custom server configuration.

    Args:
        command: Command to start the server
        name: Server name
        **kwargs: Additional configuration options

    Returns:
        MCPServerConfig instance
    """
    return MCPServerConfig(
        command=command,
        name=name,
        **kwargs,
    )


def load_servers_from_config(config: dict) -> dict[str, MCPServerConfig]:
    """
    Load server configurations from brainchain config.

    Args:
        config: MCP section of brainchain config

    Returns:
        Dict of server name to MCPServerConfig
    """
    servers = {}

    for name, server_config in config.get("servers", {}).items():
        if isinstance(server_config, dict):
            # Custom server from config
            command = server_config.get("command", [])
            if isinstance(command, str):
                command = command.split()

            servers[name] = MCPServerConfig(
                name=name,
                command=command,
                args=server_config.get("args", {}),
                env=server_config.get("env"),
                enabled=server_config.get("enabled", True),
                auto_connect=server_config.get("auto_connect", False),
                timeout=server_config.get("timeout", 30),
            )
        elif name in BUILTIN_SERVERS:
            # Enable/configure built-in server
            builtin = BUILTIN_SERVERS[name]
            servers[name] = MCPServerConfig(
                name=name,
                command=builtin.command.copy(),
                args=builtin.args.copy(),
                env=builtin.env.copy() if builtin.env else None,
                enabled=True,
                auto_connect=False,
                timeout=builtin.timeout,
            )

    return servers

File: cmd/chat/internal/mcp/servers.go
package mcp

type ServerConfig struct {
	Name        string
	Command     []string
	Args        map[string]any
	Env         map[string]string
	Enabled     bool
	AutoConnect bool
	Timeout     int
}

func (c ServerConfig) WithArgs(args map[string]any) ServerConfig {
	newArgs := make(map[string]any)
	for k, v := range c.Args {
		newArgs[k] = v
	}
	for k, v := range args {
		newArgs[k] = v
	}
	c.Args = newArgs
	return c
}

var BuiltinServers = map[string]ServerConfig{
	"filesystem": {
		Name:    "filesystem",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-filesystem"},
		Args:    map[string]any{"allowed_paths": []string{"."}},
		Enabled: true,
	},
	"fetch": {
		Name:    "fetch",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-fetch"},
		Enabled: false,
	},
	"memory": {
		Name:    "memory",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-memory"},
		Enabled: false,
	},
	"puppeteer": {
		Name:    "puppeteer",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-puppeteer"},
		Enabled: false,
	},
	"brave-search": {
		Name:    "brave-search",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-brave-search"},
		Env:     map[string]string{"BRAVE_API_KEY": ""},
		Enabled: false,
	},
	"github": {
		Name:    "github",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-github"},
		Env:     map[string]string{"GITHUB_PERSONAL_ACCESS_TOKEN": ""},
		Enabled: false,
	},
	"postgres": {
		Name:    "postgres",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-postgres"},
		Args:    map[string]any{"connection_string": ""},
		Enabled: false,
	},
	"sqlite": {
		Name:    "sqlite",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-sqlite"},
		Args:    map[string]any{"database_path": ""},
		Enabled: false,
	},
	"slack": {
		Name:    "slack",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-slack"},
		Env:     map[string]string{"SLACK_BOT_TOKEN": "", "SLACK_TEAM_ID": ""},
		Enabled: false,
	},
	"google-maps": {
		Name:    "google-maps",
		Command: []string{"npx", "-y", "@modelcontextprotocol/server-google-maps"},
		Env:     map[string]string{"GOOGLE_MAPS_API_KEY": ""},
		Enabled: false,
	},
}

func GetServerConfig(name string) *ServerConfig {
	if cfg, ok := BuiltinServers[name]; ok {
		return &cfg
	}
	return nil
}

func ListBuiltinServers() []string {
	names := make([]string, 0, len(BuiltinServers))
	for name := range BuiltinServers {
		names = append(names, name)
	}
	return names
}

func CreateServerConfig(command []string, name string, opts ...func(*ServerConfig)) ServerConfig {
	cfg := ServerConfig{
		Command: command,
		Name:    name,
		Timeout: 30,
		Enabled: true,
	}
	for _, opt := range opts {
		opt(&cfg)
	}
	return cfg
}
